### DISCUSSION TITLE
ch2-cgo/ch2-06-qsort

### DISCUSSION THREAD

**Discussion by giscus** [chai2010/advanced-go-programming-book#625] [General]
# ch2-cgo/ch2-06-qsort

Go语言高级编程

https://chai2010.cn/advanced-go-programming-book/ch2-cgo/ch2-06-qsort.html

**Comment by Future2100** [comment]
最后一段代码应该把Slice改成Sort吧

**Reply by PaiGack** [reply]
函数以及由 Sort 改成了 Slice ，所以这个地方应该是没有问题的

**Comment by flyflyhe** [comment]
package qsort

/*
#include <stdlib.h>

typedef int (*qsort_cmp_func_t)(const void* a, const void* b);
extern int _cgo_qsort_compare(void* a, void* b);
*/
import "C"
import (
	"fmt"
	"reflect"
	"runtime"
	"strconv"
	"strings"
	"sync"
	"unsafe"
)

type GoQsortCompareInfo struct {
	base     unsafe.Pointer
	elemnum  int
	elemsize int
	less     func(a, b int) bool
}

var m sync.Map

//export _cgo_qsort_compare
func _cgo_qsort_compare(a, b unsafe.Pointer) C.int {
	go_qsort_compare_infoTmp, _ := m.Load(GetGoid())
	go_qsort_compare_info := go_qsort_compare_infoTmp.(GoQsortCompareInfo)
	var (
		// array memory is locked
		base     = uintptr(go_qsort_compare_info.base)
		elemsize = uintptr(go_qsort_compare_info.elemsize)
	)

	i := int((uintptr(a) - base) / elemsize)
	j := int((uintptr(b) - base) / elemsize)

	switch {
	case go_qsort_compare_info.less(i, j): // v[i] < v[j]
		return -1
	case go_qsort_compare_info.less(j, i): // v[i] > v[j]
		return +1
	default:
		return 0
	}
}

func Slice(slice interface{}, less func(a, b int) bool) {
	sv := reflect.ValueOf(slice)
	if sv.Kind() != reflect.Slice {
		panic(fmt.Sprintf("qsort called with non-slice value of type %T", slice))
	}
	if sv.Len() == 0 {
		return
	}

	go_qsort_compare_info := GoQsortCompareInfo{}

	// baseMem = unsafe.Pointer(sv.Index(0).Addr().Pointer())
	// baseMem maybe moved, so must saved after call C.fn
	go_qsort_compare_info.base = unsafe.Pointer(sv.Index(0).Addr().Pointer())
	go_qsort_compare_info.elemnum = sv.Len()
	go_qsort_compare_info.elemsize = int(sv.Type().Elem().Size())
	go_qsort_compare_info.less = less

	m.Store(GetGoid(), go_qsort_compare_info)
	defer m.Delete(GetGoid())
	C.qsort(
		go_qsort_compare_info.base,
		C.size_t(go_qsort_compare_info.elemnum),
		C.size_t(go_qsort_compare_info.elemsize),
		C.qsort_cmp_func_t(C._cgo_qsort_compare),
	)
}

func GetGoid() int64 {
	var (
		buf [64]byte
		n   = runtime.Stack(buf[:], false)
		stk = strings.TrimPrefix(string(buf[:n]), "goroutine")
	)

	idField := strings.Fields(stk)[0]
	id, err := strconv.Atoi(idField)
	if err != nil {
		panic(fmt.Errorf("can not get goroutine id: %v", err))
	}

	return int64(id)
}
加了个sync.map 并发调用通过协程id并发调用  一时想不到其他的办法。。
